python/lsst/verify/compatibility/metricTask.py

```diff
@@ -0,0 +1,196 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import abc
+
+import lsst.pex.config
+import lsst.pipe.base as pipeBase
+
+__all__ = ["MetricTask"]
+
+
+class MetricTask(pipeBase.Task, metaclass=abc.ABCMeta):
+    """A ``MetricTask`` object computes exactly one metric, based on
+    arbitrary input datasets.
+
+    Both the metric and the input data are configurable. Metrics may be
+    associated with a data ID at any level of granularity, including
+    repository-wide. MetricTasks should assume they take multiple input
+    datasets, regardless of the expected metric granularity, unless it is
+    impossible for more than one copy of a dataset to exist.
+
+    All input data shall be considered optional in order to maximize the
+    metric's usefulness for incomplete pipeline runs or runs with optional
+    processing steps. If a metric cannot be calculated because the necessary
+    inputs are missing, the ``MetricTask`` must return `None` in place of
+    the measurement.
+
+    .. note::
+        The API is designed to make it easy to convert all ``MetricTasks`` to
+        `~lsst.pipe.base.PipelineTask` later, but this class is *not* a
+        `~lsst.pipe.base.PipelineTask` and will not work with activators,
+        quanta, or `lsst.daf.butler`.
+
+    Parameters
+    ----------
+    Same as for `lsst.pipe.base.PipelineTask`.
+    """
+
+    # There may be a specialized MetricTaskConfig later, details TBD
+    ConfigClass = lsst.pex.config.Config
+
+    def adaptArgsAndRun(self, inputData, inputDataIds, outputDataId):
+        """Compute a metric from in-memory data.
+
+        This implementation calls `run` on the contents of ``inputData``,
+        followed by calling `addStandardMetadata` on the result before
+        returning it. Any subclass that overrides this method must also call
+        `addStandardMetadata` on its measurement before returning it.
+
+        Parameters
+        ----------
+        inputData : `dict` from `str` to any
+            Dictionary whose keys are the names of input parameters and values
+            are Python-domain data objects (or lists of objects) retrieved
+            from data butler. Accepting lists of objects is strongly
+            recommended, to allow metrics to vary their granularity up to the
+            granularity of the input data without the need for extensive
+            code changes. Input objects may be `None` to represent
+            missing data.
+        inputDataIds : `dict` from `str` to `list` of dataId
+            Dictionary whose keys are the names of input parameters and values
+            are DataIds (or lists of DataIds) that task consumes for
+            corresponding dataset type. DataIds are guaranteed to match data
+            objects in ``inputData``
+        outputDataId : `dict` from `str` to dataId
```

Comment:
The problem is that AFAIK `dataId` is a concept, not a formal type. For example, it could be a `dict`, a `daf.persistence.DataId`, a `daf.butler` type that the Gen 3 group keeps waffling on, etc.. Do you have a specific alternative (and might this be a good addition to the style guide, given that we use data IDs almost as often as we use `Struct`)?

Reply:
I ended up keeping `dataId` as the value type of `inputDataIds` and `ouputDataIds`, and replacing "DataId" -> "data ID" in the description of `inputDataIds`. This means the documentation no longer exactly matches that for `PipelineTask`, but it gives some separation between `dataId` (a Python object that may be one of at least three types) and "data ID" (the information represented by a `dataId`). I don't know if `PipelineTask` wants to make the same distinction or not.

Reply:
:+1:

I can't say with any confidence that we do it consistently, but at least some daf_butler and PipelineTask API docs already use that convention.

Reply:
Can someone explain what's meant by "the information represented by a dataId" as being a differentiation to `dataId`.

This use of `dataId` is still alerting my spidey-sense to a documentation anti pattern to be honest.

Reply:
By "information" I meant that we identify datasets by things like their visit and CCD numbers, whether or not we put those numbers into a particular object. Kind of how like a `float` represents a number, but "number" is a mathematical concept that exists whether or not you program with `float`s.

Since the discussion on `#dm-hackweek` quickly converged on using "data ID" as the type description and not just in the explanatory text, I think the distinction is moot now.

Reply:
Ok, thanks, I wasn't sure it was resolved.